compiler/rustc_mir_transform/src/coverage/repair.rs

```diff
@@ -0,0 +1,58 @@
+use rustc_middle::mir::coverage::{CounterId, CoverageKind};
+use rustc_middle::mir::{
+    self, Coverage, MirPass, SourceInfo, Statement, StatementKind, START_BLOCK,
+};
+use rustc_middle::ty::TyCtxt;
+use rustc_span::DUMMY_SP;
+
+/// If a function has been [instrumented for coverage](super::InstrumentCoverage),
+/// but MIR optimizations subsequently remove all of its [`CoverageKind::CounterIncrement`]
+/// statements (e.g. because bb0 is unreachable), then we won't generate any
+/// `llvm.instrprof.increment` intrinsics. LLVM will think the function is not
+/// instrumented, and it will disappear from coverage mappings and coverage reports.
+///
+/// This pass detects when that has happened, and re-inserts a dummy counter-increment
+/// statement so that LLVM knows to treat the function as instrumented.
```

Comment:
I wonder if this should be done by codegen directly. When starting codegen for a mono-item, check if `function_coverage_info` is set, and add this logic?

Reply:
Yeah, that sounds reasonable. I'll have to track down the right place to add a hook, but once I do the result should be more self-contained in codegen.